README.md

```diff
@@ -362,6 +361,195 @@ async fn main() -> PyResult<()> {
+
+One clear disadvantage to this approach (besides the verbose naming) is that the Rust application has to explicitly track its references to the Python event loop. In native libraries, we can't make any assumptions about the underlying event loop, so the only reliable way to make sure our conversions work properly is to store a reference to the current event loop to use later on.
+
+```rust
+use pyo3::prelude::*;
+
+#[pyfunction]
+fn sleep(py: Python) -> PyResult<&PyAny> {
+    let current_loop = pyo3_asyncio::get_running_loop(py)?;
+    let loop_ref = PyObject::from(current_loop);
+
+    // Convert the async move { } block to a Python awaitable
+    pyo3_asyncio::tokio::future_into_py_with_loop(current_loop, async move {
+        let py_sleep = Python::with_gil(|py| {
+            // Sometimes we need to call other async Python functions within
+            // this future. In order for this to work, we need to track the 
+            // event loop from earlier.
+            pyo3_asyncio::into_future_with_loop(
+                loop_ref.as_ref(py), 
+                py.import("asyncio")?.call_method1("sleep", (1,))?
+            )
+        })?;
+
+        py_sleep.await?;
+
+        Ok(Python::with_gil(|py| py.None()))
+    })
+}
+
+#[pymodule]
+fn my_mod(py: Python, m: &PyModule) -> PyResult<()> {
+    m.add_function(wrap_pyfunction!(sleep, m)?)?;
+    Ok(())
+}
+```
+
+> A naive solution to this tracking problem would be to cache a global reference to the asyncio event loop that all PyO3 Asyncio conversions can use. In fact this is what we did in PyO3 Asyncio `v0.13`. This works well for applications, but it soon became clear that this is not so ideal for libraries. Libraries usually have no direct control over how the event loop is managed, they're just expected to work with any event loop at any point in the application. This problem is compounded further when multiple event loops are used in the application since the global reference will only point to one.
+
+Another disadvantage to this explicit approach that is less obvious is that we can no longer call our `#[pyfunction] fn sleep` on a Rust runtime since `asyncio.get_running_loop` only works on Python threads! It's clear that we need a slightly more flexible approach.
+
+In order to detect the Python event loop at the callsite, we need something like `asyncio.get_running_loop` that works for _both Python and Rust_. In Python, `asyncio.get_running_loop` uses thread-local data to retrieve the event loop associated with the current thread. What we need in Rust is something that can retrieve the Python event loop associated with the current _task_.
+
+Enter `pyo3_asyncio::<runtime>::get_current_loop`. This function first checks task-local data for a Python event loop, then falls back on `asyncio.get_running_loop` if no task-local event loop is found. This way both bases are convered.
```

Comment:
Random, but this also made me realise that most of my concerns with threadedness, and event loops on multiple threads, will be covered.

There's still 1 edgecase, which is if someone somehow spawned a future that escaped the task scope, or introduced one which hasn't been assigned from pyo3-asyncio at all (from a control flow that never spawned from python, i.e. built-in python). But i think this is good enough for 95% of all usage cases, and so 99.99% of all usages

Reply:
Yeah, it might be good to make a variant of `get_current_loop` that doesn't panic (something like `try_get_current_loop`). I guess the main question surrounding this is if there is ever a case where this is a recoverable error rather than a design flaw in the user's code. I agree that this is something that can be addressed later on in a follow-up though.